### PR TITLE
Allow SubmitCGAP to work for Fourfront's new ingestion facility (C4-996)

### PR DESCRIPTION
**Pull request by netsettler**
Implementation of [C4-996](https://hms-dbmi.atlassian.net/browse/C4-996), creating support for using the `SubmitCGAP` library to talk to the new `Fourfront` ingestion facility that David is porting from `cgap-portal`.

* Adds a `submit-ontology` command

* In `base.py`:

  *  New class `GenericKeyManager` to manage possibility of either `Fourfront` or `CGAP` keys files (`~/.fourfront-keys.json` vs `~/.cgap-keys.json`)

* Appropriate unit test coverage.

[C4-996]: https://hms-dbmi.atlassian.net/browse/C4-996?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: submit_cgap/scripts/submit_ontology.py
Comment (willronchetti) [inline submit_cgap/scripts/submit_ontology.py:13]: Lifted from Gene list submission

File: submit_cgap/scripts/submit_ontology.py
Comment (willronchetti) [inline submit_cgap/scripts/submit_ontology.py:17]: I think this can be a option if the user wants to generate the terms locally but this could be a large file, we might consider instead (or additionally) allowing the generation of this file in the ingester.

**Review by willronchetti — APPROVED** [review]
Generally looks good just a few small comments and suggestions for future proofing this no matter the strategy we go with internally

File: submit_cgap/scripts/submit_ontology.py
Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:32]: I assumed it was a permissioning issue, that it might be needed. I get that an admin might not need it, but is this command only intended to be runnable by admins? SubmitCGAP can't know who is/isn't. If it's possible that someone's role might give them permission to also run this, then it should at least have it as an option.

Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:32]: @aschroed You probably understand the  permissioning here, so maybe you can help clarify that. (I'm as interested in intended permissioning as implemented permissioning, since those might diverge as David is implementing this in parallel and not quite done.)

File: submit_cgap/submission.py
Comment (netsettler) [reply submit_cgap/submission.py:179]: The code is right, but I've now fixed the comment. (Locally. To be uploaded shortly.)

File: submit_cgap/scripts/submit_ontology.py
Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:13]: Also fixed locally. To be uploaded shortly.

File: submit_cgap/scripts/submit_ontology.py
Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:17]: From what would the ontology be generated? Doesn't it have to have input from the user somehow? I think I'm missing something important here.

(For now the intent was to just have some way of doing submitting, so David can test things, but you're right that if merged it would be user-visible. On the other hand, if this gets complicated, I assume we could evolve this once done to implement other use cases.)

Comment (willronchetti) [reply submit_cgap/scripts/submit_ontology.py:17]: The `generate_ontology` command just takes an ontology type alone (or multiple) and will generate the file. We could have the command operate this way as well. See https://github.com/4dn-dcic/fourfront/blob/master/src/encoded/commands/generate_ontology.py

File: submit_cgap/submission.py
Comment (dmichaels-harvard) [inline submit_cgap/submission.py:215]: Typo introduced somehow here, broke CGAP specific submit, fixed and checked in myself, hope you don't mind.

File: submit_cgap/scripts/submit_ontology.py
Comment (aschroed) [reply submit_cgap/scripts/submit_ontology.py:32]: Ontology updates are only by admin or the system.  The enumerated available status values for this Item makes them uneditable by non-admin in any case.  Currently, the submitter properties show that the 'load_xl' user of the system submits and updates terms.   OntologyTerm schema lacks lab and award properties and therefore there is not 'attribution' tab for this item type. 

The original intent was to have this all completely automated as part of the system.  It has subsequently shown that automatic updates are not really possible and that manual review of the terms generated by the script for updates is generally needed for a variety of reasons usually out of our control.

That's why I am not sure if @willronchetti 's suggestion about incorporating generate_ontology into the ingestion per se will work.

Intended permissioning is to have tight control of who can submit or change ontology terms - particularly those added via processing of files/terms provided by the official ontology maintainers.  We have our own 4DN controlled vocabulary ontology that we are a bit more flexible with where we can add terms we need that don't exist in existing ontologies yet.  That said I am not against more flexibility and being able to identify the specific person that did an update could be useful though attribution for these items doesn't add much IMO.

File: submit_cgap/submission.py
Comment (netsettler) [reply submit_cgap/submission.py]: (You fixed this.)

File: submit_cgap/scripts/submit_ontology.py
Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:17]: You misunderstood my concern. I'm sure you're right about the command. But in order to have permission to run the command by invocation from the outside, I assume the lab and award are potentially used for computing the permissions object. Maybe that's already factored in based on who the person is, and nothing is added by saying "I'm acting in this role", though honestly I think it would be better for hospital business rules if you had to (at least in CGAP) say which role you were acting in, since it creates a firewall on what information an individual case should see, and since a person wearing multiple hats should not necessarily be able to see all the things at once that they are authorized to see separately. I guess we could fix that later, though.

File: submit_cgap/scripts/submit_ontology.py
Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:32]: @aschroed Thanks for that info. So it sounds like we could add a role that a person had to have (as an alternative to admin) in order to be able to perform this action. But to be clear, do you think there's value requiring either of Lab and Award. They're easily enough made optional, but Award, at least, is hard to reconstruct after-the-fact, so I don't want to miss the chance of recoring things we might later need, and I can't evaluate that very well. Is this action ever award-specific? It sounds like not. Is the role ever award-specific? I _think_ it sounds like not, but independent of the philosophizing, those are my concrete needs at this point: just to know if we should require either of those, too. Later we could extend permissioning, and for now we could make the submission only get accepted for processing under if you're an admin. Does that sound right?

Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:32]: @aschroed Also, though there ia secondary question of whether we shold process the review in two steps. We could start with a single step and then break it up, or we could hold this functionality until we have a two-step process. We could offer some way to review the intermediate results before submitting them. That would be unusual but not impossible, and then there could be a secondary step that commits the result. It might require a UI step for the review, or perhaps we could make a spreadsheet of information for download that the person could then do an approval action on. It's possible these items are better discussed on the ticket than on the PR, but it's OK, I'll move relevant parts of this discussion later. Since it started here, we might as well finish it out here.

File: submit_cgap/scripts/submit_ontology.py
Comment (netsettler) [reply submit_cgap/scripts/submit_ontology.py:17]: @aschroed you may have thoughts here as well, or this may be redundant with the other discussion. I _think_ it's @willronchetti I'm waiting on here, but he may be waiting to see your replies to the other stuff above. It's hard to know in cases like this if I'm assembling all info togetther but I wanted to keep this moving along so we can get some resolution.

File: submit_cgap/scripts/submit_ontology.py
Comment (aschroed) [reply submit_cgap/scripts/submit_ontology.py:32]: > So it sounds like we could add a role that a person had to have (as an alternative to admin) in order to be able to perform this action.

That would be nice - similar to how a 'submits_for' works in Submit4DN though of course since that is tied to Lab in the authorization step really not that similar given my next answer.

> for now we could make the submission only get accepted for processing under if you're an admin. Does that sound right?

Yes - at least for now.

> We could start with a single step and then break it up, or we could hold this functionality until we have a two-step process. We could offer some way to review the intermediate results before submitting them. That would be unusual but not impossible, and then there could be a secondary step that commits the result. 

It would be quite dangerous to do this in a single step without review.  Currently there are 2 separate scripts - one to generate the json to load and the other to load it.  I recently found a 'bug' in the generation script that was due to script updates that did not anticipate a particular non-common scenario.  And the ontology maintainers themselves make changes or add bugs that are independent of our code.  There is some validation in the generator script but what action is required when things pop up is not always straightforward.

I am not familiar with exactly how the ingester works but for practicality for 4DN ontology terms and I don't see this need be much different for SMAHT ontologies I think the term generation should remain separate from the loading - the generated file can be presumably uploaded to AWS manually if needed for subsequent ingestion.  Only commenting on dealing with ontology term management here.

**Comment by netsettler** [comment]
@aschroed, @willronchetti, and @dmichaels-harvard, I'd kind of like to merge this functionality and yet it seems not to do precisely what we want and I'm not wanting to release functionality to users that would be premature and confusing.

Would it be acceptable if I just disabled the command setup in pyproject.toml but left the rest of the code as-is, and then in a future PR we could turn this command on, or something similar? For now the functionality would be "parked" but there would not be a big PR pending. Then I could start work on the next ticket from master rather than building further on this.

**Comment by netsettler** [comment]
@dmichaels-harvard Under that proposed solution, I could set it up so you could still invoke the script as an executable (or just by invoking it as `python` with a script argument). I don't want to take away functionality you're using for debugging, but it would just require a slightly different syntax while we figure out the right real interface to advertise.

File: submit_cgap/submission.py
Comment (willronchetti) [inline submit_cgap/submission.py:187]: Given Andy's comments it's not clear to me failure behavior is desired here.

File: submit_cgap/submission.py
Comment (willronchetti) [inline submit_cgap/submission.py:208]: Same comment as for `award` above.

File: submit_cgap/submission.py
Comment (willronchetti) [inline submit_cgap/submission.py:371]: I contest these are required in all cases, unlike for CGAP, and if the loadxl ingestion is ported back to CGAP the same will be true there

**Review by dmichaels-harvard — APPROVED** [review]

**Pull request merged**
